src/XrdOssStats/XrdStatsFileSystem.hh

```diff
@@ -0,0 +1,128 @@
+    void      Disc(XrdOucEnv &env) override;
+    void      EnvInfo(XrdOucEnv *env) override;
+    uint64_t  Features() override;
+    int       FSctl(int cmd, int alen, const char *args, char **resp=0) override;
+    int       Init(XrdSysLogger *lp, const char *cfn) override;
+    int       Init(XrdSysLogger *lp, const char *cfn, XrdOucEnv *env) override;
+    int       Mkdir(const char *path, mode_t mode, int mkpath=0,
+                        XrdOucEnv  *env=0) override;
+    int       Reloc(const char *tident, const char *path,
+                        const char *cgName, const char *anchor=0) override;
+    int       Remdir(const char *path, int Opts=0, XrdOucEnv *env=0) override;
+    int       Rename(const char *oPath, const char *nPath,
+                         XrdOucEnv  *oEnvP=0, XrdOucEnv *nEnvP=0) override;
+    int       Stat(const char *path, struct stat *buff,
+                       int opts=0, XrdOucEnv *env=0) override;
+    int       Stats(char *buff, int blen) override;
+    int       StatFS(const char *path, char *buff, int &blen,
+                         XrdOucEnv  *env=0) override;
+    int       StatLS(XrdOucEnv &env, const char *path,
+                         char *buff, int &blen) override;
+    int       StatPF(const char *path, struct stat *buff, int opts) override;
+    int       StatPF(const char *path, struct stat *buff) override;
+    int       StatVS(XrdOssVSInfo *vsP, const char *sname=0, int updt=0) override;
+    int       StatXA(const char *path, char *buff, int &blen,
+                         XrdOucEnv *env=0) override;
+    int       StatXP(const char *path, unsigned long long &attr,
+                         XrdOucEnv  *env=0) override;
+    int       Truncate(const char *path, unsigned long long fsize,
+                           XrdOucEnv *env=0) override;
+    int       Unlink(const char *path, int Opts=0, XrdOucEnv *env=0) override;
+    int       Lfn2Pfn(const char *Path, char *buff, int blen) override;
+    const char       *Lfn2Pfn(const char *Path, char *buff, int blen, int &rc) override;
+
+private:
+    static void * AggregateBootstrap(void *instance);
+    void AggregateStats();
+
+    XrdXrootdGStream* m_gstream{nullptr};
+
+    XrdOss *m_oss;
+    XrdOucEnv *m_env;
+    XrdSysError m_log;
+
+    class OpTimer {
+        public:
+            OpTimer(std::atomic<uint64_t> &op_count, std::atomic<uint64_t> &slow_op_count, std::atomic<uint64_t> &timing, std::atomic<uint64_t> &slow_timing, std::chrono::steady_clock::duration duration);
+            ~OpTimer();
+
+        private:
+            std::atomic<uint64_t> &m_op_count;
+            std::atomic<uint64_t> &m_slow_op_count;
+            std::atomic<uint64_t> &m_timing;
+            std::atomic<uint64_t> &m_slow_timing;
+            std::chrono::steady_clock::time_point m_start;
+            std::chrono::steady_clock::duration m_slow_duration;
+    };
+
+    struct OpRecord {
+        std::atomic<uint64_t> m_read_ops{0};
```

Comment:
OK, I understand why you may want to intercept to get timing data which we don't do. But much of this information (i.e. counts) is already reported in the f-stream. So, now we have a duplicate way of doing it and potentiually collecting the majority of the same statistics (and reporting them) twice. So, what's the actual motivation here?

Frankly, if this is a preferred way of doing things now then I would make oss gStream incompatible with f-stream you can do one or the other but not both.

Reply:
Is this a complete overlap though?

My understanding is the f-stream works at the xrootd protocol handler level; this works at the OSS level.  The OSS level, for example, is used directly by XrdPfc and XrdTpc, meaning that this is seeing many reads that aren't reflected in the f-stream.